clients/cobol-lsp-vscode-extension/src/__tests__/services/copybook/E4ECopybookService.spec.ts

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (c) 2024 Broadcom.
+ * The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Broadcom, Inc. - initial API and implementation
+ */
+import * as vscode from "vscode";
+import { E4ECopybookService } from "../../../services/copybook/E4ECopybookService";
+import { e4eMock } from "./download/getE4EMock.utility";
+import { Utils } from "../../../services/util/Utils";
+
+describe("e4e copybook service tests", () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+  it("checks valid e4e api retrieved", async () => {
+    const ext = { exports: { ...e4eMock }, activate: jest.fn() };
+    vscode.extensions.getExtension = jest.fn().mockReturnValue(ext);
+
+    const spyValidate = jest.spyOn(Utils, "validateE4E");
+    const api = await E4ECopybookService.getE4EAPI();
+    expect(spyValidate).toHaveBeenCalled();
+    expect(api === e4eMock);
+  });
+
+  it("check getE4EClient assembles client correctly / check getE4EClient returns already assembled client once called with same Uri ", async () => {
+    E4ECopybookService.getE4EAPI = jest.fn().mockReturnValue(e4eMock);
+    const res = await E4ECopybookService.getE4EClient("document-uri");
+    expect(res?.elements).toEqual(res?.elements);
+    expect(res?.members).toEqual(res?.members);
+    const spyApi = jest.spyOn(E4ECopybookService, "getE4EAPI");
+    expect(spyApi).toHaveBeenCalledTimes(1);
+    await E4ECopybookService.getE4EClient("document-uri");
+    expect(spyApi).toHaveBeenCalledTimes(1);
+    await E4ECopybookService.getE4EClient("new-uri");
+    expect(spyApi).toHaveBeenCalledTimes(2);
```

Comment:
What is this testing? Why is it so important to make sure the `getE4EAPI` gets called? This seems like a test for an implementation detail that will just make any future refactoring harder.

Reply:
it is checking assembled client for a document is re assembles or gets the resolved one.

Reply:
Calling `getE4EAPI` is at best an inefficiency of the current implementation, once obtained, the API does not change. Making the implementation more efficient would break this test.